File: Sources/Heeler/Transport/SSHChannelAdmission.swift
import Foundation
import Synchronization

/// Separates OpenSSH forwarding and session admission while enforcing one
/// finite ceiling over every live SSH channel on a Host connection.
///
/// Session categories leave headroom under sshd's default `MaxSessions` of 10
/// (ADR 0011). That server-side limit is the real binding constraint for exec,
/// PTY, and SFTP channels, and the app cannot observe it: an abandoned channel
/// teardown still releases the app admission lease unconditionally, while the
/// server session slot can remain occupied (#148). App counters and
/// `isConnected` may therefore still look available after the server has
/// already spent the slot; they cannot promise MaxSessions safety.
actor SSHChannelAdmission {
    enum ChannelClass: Sendable {
        case ordinaryForwarding
        case events
        case ordinarySession
        case attach
    }

    struct Limits: Sendable, Equatable {
        let ordinaryForwarding: Int
        let events: Int
        let ordinarySession: Int
        let attach: Int
        let connection: Int

        /// Production budgets. `connection` restates the sum of the four
        /// category budgets (8+1+8+1), so the connection-level guard does not
        /// bind under these categories — it only constrains a future category
        /// that would otherwise grow without a connection cap. Per-category
        /// budgets are the live admission limits today; do not treat 18 as a
        /// tighter guarantee than those budgets or than server `MaxSessions`.
        static let production = Limits(
            ordinaryForwarding: 8,
            events: 1,
            ordinarySession: 8,
            attach: 1,
            connection: 18)
    }

    struct Snapshot: Sendable, Equatable {
        let ordinaryForwarding: Int
        let events: Int
        let ordinarySession: Int
        let attach: Int
        let connection: Int
    }

    private struct Waiter {
        let id: UInt64
        let channelClass: ChannelClass
        let continuation: CheckedContinuation<Void, any Error>
    }

    private let limits: Limits
    private var counts = Snapshot(
        ordinaryForwarding: 0,
        events: 0,
        ordinarySession: 0,
        attach: 0,
        connection: 0)
    private var waiters: [Waiter] = []
    private var pendingRequests: Set<UInt64> = []
    private var cancelledRequests: Set<UInt64> = []
    private var nextRequestID: UInt64 = 0

    init(limits: Limits = .production) {
        precondition(limits.ordinaryForwarding > 0)
        precondition(limits.events > 0)
        precondition(limits.ordinarySession > 0)
        precondition(limits.attach > 0)
        precondition(limits.connection > 0)
        self.limits = limits
    }

    func withChannel<Value: Sendable>(
        _ channelClass: ChannelClass,
        operation: @escaping @Sendable () async throws -> Value
    ) async throws -> Value {
        let lease = try await acquire(channelClass)
        do {
            let value = try await operation()
            await lease.release()
            return value
        } catch {
            await lease.release()
            throw error
        }
    }

    func acquire(_ channelClass: ChannelClass) async throws -> SSHChannelAdmissionLease {
        try Task.checkCancellation()
        nextRequestID &+= 1
        let requestID = nextRequestID
        pendingRequests.insert(requestID)
        defer {
            pendingRequests.remove(requestID)
            cancelledRequests.remove(requestID)
        }

        try await withTaskCancellationHandler {
            try await withCheckedThrowingContinuation {
                (continuation: CheckedContinuation<Void, any Error>) in
                if cancelledRequests.contains(requestID) {
                    continuation.resume(throwing: CancellationError())
                } else if canAcquire(channelClass) {
                    increment(channelClass)
                    continuation.resume()
                } else {
                    waiters.append(Waiter(
                        id: requestID,
                        channelClass: channelClass,
                        continuation: continuation))
                }
            }
        } onCancel: {
            Task { await self.cancel(requestID: requestID) }
        }

        if Task.isCancelled {
            release(channelClass)
            throw CancellationError()
        }
        return SSHChannelAdmissionLease(admission: self, channelClass: channelClass)
    }

    func snapshot() -> Snapshot { counts }

    fileprivate func release(_ channelClass: ChannelClass) {
        decrement(channelClass)
        resumeEligibleWaiters()
    }

    private func cancel(requestID: UInt64) {
        guard pendingRequests.contains(requestID) else { return }
        if let index = waiters.firstIndex(where: { $0.id == requestID }) {
            waiters.remove(at: index).continuation.resume(throwing: CancellationError())
        } else {
            cancelledRequests.insert(requestID)
        }
    }

    private func resumeEligibleWaiters() {
        while let index = waiters.firstIndex(where: { canAcquire($0.channelClass) }) {
            let waiter = waiters.remove(at: index)
            increment(waiter.channelClass)
            waiter.continuation.resume()
        }
    }

    private func canAcquire(_ channelClass: ChannelClass) -> Bool {
        guard counts.connection < limits.connection else { return false }
        switch channelClass {
        case .ordinaryForwarding:
            return counts.ordinaryForwarding < limits.ordinaryForwarding
        case .events:
            return counts.events < limits.events
        case .ordinarySession:
            return counts.ordinarySession < limits.ordinarySession
        case .attach:
            return counts.attach < limits.attach
        }
    }

    private func increment(_ channelClass: ChannelClass) {
        counts = Snapshot(
            ordinaryForwarding: counts.ordinaryForwarding
                + (channelClass == .ordinaryForwarding ? 1 : 0),
            events: counts.events + (channelClass == .events ? 1 : 0),
            ordinarySession: counts.ordinarySession
                + (channelClass == .ordinarySession ? 1 : 0),
            attach: counts.attach + (channelClass == .attach ? 1 : 0),
            connection: counts.connection + 1)
    }

    private func decrement(_ channelClass: ChannelClass) {
        precondition(counts.connection > 0)
        counts = Snapshot(
            ordinaryForwarding: counts.ordinaryForwarding
                - (channelClass == .ordinaryForwarding ? 1 : 0),
            events: counts.events - (channelClass == .events ? 1 : 0),
            ordinarySession: counts.ordinarySession
                - (channelClass == .ordinarySession ? 1 : 0),
            attach: counts.attach - (channelClass == .attach ? 1 : 0),
            connection: counts.connection - 1)
    }
}

final class SSHChannelAdmissionLease: Sendable {
    private let admission: SSHChannelAdmission
    private let channelClass: SSHChannelAdmission.ChannelClass
    private let released = Mutex(false)

    fileprivate init(
        admission: SSHChannelAdmission,
        channelClass: SSHChannelAdmission.ChannelClass
    ) {
        self.admission = admission
        self.channelClass = channelClass
    }

    func release() async {
        let shouldRelease = released.withLock { released in
            guard !released else { return false }
            released = true
            return true
        }
        if shouldRelease { await admission.release(channelClass) }
    }

    deinit {
        let admission = admission
        let channelClass = channelClass
        let shouldRelease = released.withLock { released in
            guard !released else { return false }
            released = true
            return true
        }
        if shouldRelease {
            Task { await admission.release(channelClass) }
        }
    }
}

File: android/core/src/main/java/dev/bybee/heeler/core/transport/SshChannelAdmission.kt
package dev.bybee.heeler.core.transport

import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Separates OpenSSH forwarding and session admission while enforcing one
 * finite ceiling over every live SSH channel on a Host connection.
 *
 * Session categories leave headroom under sshd's default `MaxSessions` of 10
 * (ADR 0011). That server-side limit is the real binding constraint for exec,
 * PTY, and SFTP channels, and the app cannot observe it: an abandoned channel
 * teardown still releases the app admission lease unconditionally, while the
 * server session slot can remain occupied. App counters and `isConnected` may
 * therefore still look available after the server has spent the slot; they
 * cannot promise MaxSessions safety.
 */
class SshChannelAdmission(
    private val limits: Limits = Limits.PRODUCTION,
) {
    enum class ChannelClass {
        ORDINARY_FORWARDING,
        EVENTS,
        ORDINARY_SESSION,
        ATTACH,
    }

    data class Limits(
        val ordinaryForwarding: Int,
        val events: Int,
        val ordinarySession: Int,
        val attach: Int,
        val connection: Int,
    ) {
        init {
            require(ordinaryForwarding > 0)
            require(events > 0)
            require(ordinarySession > 0)
            require(attach > 0)
            require(connection > 0)
        }

        companion object {
            /**
             * Production budgets. `connection` restates the sum of the four
             * category budgets (8+1+8+1), so the connection-level guard does
             * not bind under these categories. Per-category budgets are the
             * admission limits; do not treat 18 as a tighter guarantee than
             * them or than server `MaxSessions`.
             */
            val PRODUCTION = Limits(
                ordinaryForwarding = 8,
                events = 1,
                ordinarySession = 8,
                attach = 1,
                connection = 18,
            )
        }
    }

    data class Snapshot(
        val ordinaryForwarding: Int,
        val events: Int,
        val ordinarySession: Int,
        val attach: Int,
        val connection: Int,
    )

    private data class Waiter(
        val channelClass: ChannelClass,
        val signal: CompletableDeferred<Unit>,
        var admitted: Boolean = false,
    )

    private val mutex = Mutex()
    private var counts = Snapshot(0, 0, 0, 0, 0)
    private val waiters = ArrayDeque<Waiter>()

    suspend fun <T> withChannel(
        channelClass: ChannelClass,
        operation: suspend () -> T,
    ): T {
        val lease = acquire(channelClass)
        return try {
            operation()
        } finally {
            lease.release()
        }
    }

    suspend fun acquire(channelClass: ChannelClass): SshChannelAdmissionLease {
        val waiter = mutex.withLock {
            if (canAcquire(channelClass)) {
                increment(channelClass)
                null
            } else {
                Waiter(channelClass, CompletableDeferred<Unit>()).also(waiters::addLast)
            }
        }
        if (waiter != null) {
            try {
                waiter.signal.await()
            } catch (failure: CancellationException) {
                val releaseLease = mutex.withLock {
                    if (waiters.remove(waiter)) {
                        false
                    } else {
                        waiter.admitted
                    }
                }
                if (releaseLease) release(channelClass)
                throw failure
            }
        }
        return SshChannelAdmissionLease(this, channelClass)
    }

    suspend fun snapshot(): Snapshot = mutex.withLock { counts }

    private suspend fun release(channelClass: ChannelClass) {
        mutex.withLock {
            decrement(channelClass)
            resumeEligibleWaiters()
        }
    }

    private fun resumeEligibleWaiters() {
        for (waiter in waiters.toList()) {
            if (!canAcquire(waiter.channelClass)) continue
            waiters.remove(waiter)
            increment(waiter.channelClass)
            waiter.admitted = true
            waiter.signal.complete(Unit)
        }
    }

    private fun canAcquire(channelClass: ChannelClass): Boolean {
        if (counts.connection >= limits.connection) return false
        return when (channelClass) {
            ChannelClass.ORDINARY_FORWARDING -> counts.ordinaryForwarding < limits.ordinaryForwarding
            ChannelClass.EVENTS -> counts.events < limits.events
            ChannelClass.ORDINARY_SESSION -> counts.ordinarySession < limits.ordinarySession
            ChannelClass.ATTACH -> counts.attach < limits.attach
        }
    }

    private fun increment(channelClass: ChannelClass) {
        counts = counts.copy(
            ordinaryForwarding = counts.ordinaryForwarding +
                if (channelClass == ChannelClass.ORDINARY_FORWARDING) 1 else 0,
            events = counts.events + if (channelClass == ChannelClass.EVENTS) 1 else 0,
            ordinarySession = counts.ordinarySession +
                if (channelClass == ChannelClass.ORDINARY_SESSION) 1 else 0,
            attach = counts.attach + if (channelClass == ChannelClass.ATTACH) 1 else 0,
            connection = counts.connection + 1,
        )
    }

    private fun decrement(channelClass: ChannelClass) {
        check(counts.connection > 0)
        counts = counts.copy(
            ordinaryForwarding = counts.ordinaryForwarding -
                if (channelClass == ChannelClass.ORDINARY_FORWARDING) 1 else 0,
            events = counts.events - if (channelClass == ChannelClass.EVENTS) 1 else 0,
            ordinarySession = counts.ordinarySession -
                if (channelClass == ChannelClass.ORDINARY_SESSION) 1 else 0,
            attach = counts.attach - if (channelClass == ChannelClass.ATTACH) 1 else 0,
            connection = counts.connection - 1,
        )
    }

    class SshChannelAdmissionLease internal constructor(
        private val admission: SshChannelAdmission,
        private val channelClass: ChannelClass,
    ) {
        private val released = AtomicBoolean(false)

        suspend fun release() {
            if (released.compareAndSet(false, true)) admission.release(channelClass)
        }
    }
}
